WordDistanceInSpark.scala

```diff
@@ -1 +1,241 @@
+  at org.apache.spark.ContextCleaner$$anonfun$org$apache$spark$ContextCleaner$$keepCleaning$1.apply$mcV$sp(ContextCleaner.scala:146)
+  at org.apache.spark.ContextCleaner$$anonfun$org$apache$spark$ContextCleaner$$keepCleaning$1.apply(ContextCleaner.scala:144)
+  at org.apache.spark.ContextCleaner$$anonfun$org$apache$spark$ContextCleaner$$keepCleaning$1.apply(ContextCleaner.scala:144)
+  at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1618)
+  at org.apache.spark.ContextCleaner.org$apache$spark$ContextCleaner$$keepCleaning(ContextCleaner.scala:143)
+  at org.apache.spark.ContextCleaner$$anon$3.run(ContextCleaner.scala:65)
+ERROR Utils: Uncaught exception in thread SparkListenerBus
+java.lang.InterruptedException
+  at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:998)
+  at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
+  at java.util.concurrent.Semaphore.acquire(Semaphore.java:312)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1$$anonfun$run$1.apply$mcV$sp(AsynchronousListenerBus.scala:62)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1$$anonfun$run$1.apply(AsynchronousListenerBus.scala:61)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1$$anonfun$run$1.apply(AsynchronousListenerBus.scala:61)
+  at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1618)
+  at org.apache.spark.util.AsynchronousListenerBus$$anon$1.run(AsynchronousListenerBus.scala:60)*/
+
+
+  
+  
+  
+  type Hit = (Long, String) // (index, the significant word found at index)
+  type SameWordInterval = ((Long,Long),String) // ( (startIdx, endIdx), significant word)
+  
+  /**
+   * Finds the distance between given words in the text, defined as 
+   * minimum amount of other words in between occurrences of the given words.
+   * 
+   * Distance between a word and itself is 0.
+   * If one of the words does not occur in the text, distance is None
+   *  
+   * @param textFilePath
+   * @param word1
+   * @param word2
+   * @param sc sparkContext
+   * @return
+   */
+  def wordDistance(textFilePath:String, word1:String, word2:String)(implicit sc:SparkContext):Option[Long] = {
+    if (word1 == word2) return Some(0)
+    if (word1==null || word2==null) return None
+    
+    // need access to a cluster to tweak this experimentally, so this is just a guess. 
+    // I don't know whether even to fiddle with this at all or just take the defaultParallelism.
+    implicit val nbPartitions = sc.defaultParallelism * 2
+    
+    // would this normally have to be broadcasted to every node then?
+    val lines:RDD[String] = sc.textFile(textFilePath, nbPartitions)
+    
+    // look at text file as sequence of words
+    val words = asFlatWordList(lines)
+    
+    // we filter out all words that are not word1 or word2
+    val orderedHits:RDD[Hit] = findHitsInOrder(words, word1, word2)
```

Comment:
Let me suggest something that you might find an improvement then.
val sortedHits:RDD[Hit] = findAndSortHitsOf(words, word1, word2)

This is of course me guessing that the confusion is about using inOrder instead of sorted?
It could also be just the concept of a Hit:  my usage of Hit = (idx, word) implies that the word is one of the two significant words, so we have a hit instead of a miss. The method filters out the misses, keeping only the hits, and returns them sorted from start of document to end of document.

Oh, how much easier it is when you're at a whiteboard together, and can do the thinking out loud :)